sim/aviron/src/lib/mcu.zig

```diff
@@ -0,0 +1,172 @@
+const std = @import("std");
+const io_mod = @import("io.zig");
+const memory = @import("memory.zig");
+const Cpu = @import("Cpu.zig");
+
+/// MCU configuration defines the memory layout and special register addresses
+/// for a specific AVR microcontroller.
+pub const Config = struct {
+    /// Human-readable name of the MCU
+    name: []const u8,
+
+    /// Flash memory size in bytes (must be 2-aligned)
+    flash_size: usize,
+
+    /// SRAM size in bytes
+    sram_size: usize,
+
+    /// SRAM base address in data space
+    sram_base: u24,
+
+    /// EEPROM size in bytes
+    eeprom_size: usize,
+
+    /// Code model (determines PC width)
+    code_model: Cpu.CodeModel,
+
+    /// Instruction set variant
+    instruction_set: Cpu.InstructionSet,
+
+    /// Special I/O register addresses
+    special_io: SpecialIoConfig,
+
+    /// Start of IO window in data address space (inclusive).
+    io_window_base: u24,
+    /// End of IO window in data address space (inclusive).
+    io_window_end: u24,
+};
+
+/// Convenience container for constructed memory spaces.
+pub const Spaces = struct {
+    data: memory.MemorySpace,
+    io: memory.MemorySpace,
+    prog: memory.MemorySpace,
+
+    pub fn deinit(self: *const Spaces, alloc: std.mem.Allocator) void {
+        self.data.deinit(alloc);
+        self.io.deinit(alloc);
+        self.prog.deinit(alloc);
+    }
+};
+
+/// Build memory spaces (data, io, program) for the given MCU configuration.
+pub fn build_spaces(
+    alloc: std.mem.Allocator,
+    cfg: Config,
+    flash: *const io_mod.Flash,
+    ram: *io_mod.RAM,
+    io_mem: *io_mod.IO,
+) !Spaces {
+    // IO window size
+    const io_size: usize = @intCast(cfg.io_window_end - cfg.io_window_base + 1);
+
+    // Data space: IO window mapped into data space (at base), then SRAM at sram_base
+    var data_seg_buf: [2]memory.Segment = undefined;
+    data_seg_buf[0] = .{ .at = cfg.io_window_base, .size = io_size, .backend = memory.Backend.fromIO(io_mem) };
+    data_seg_buf[1] = .{ .at = cfg.sram_base, .size = ram.size, .backend = memory.Backend.fromRAM(ram) };
+    const data_space = try memory.MemorySpace.init(alloc, data_seg_buf[0..]);
+
+    // IO space: IO addresses starting at 0
+    var io_seg_buf: [1]memory.Segment = undefined;
+    io_seg_buf[0] = .{ .at = 0, .size = io_size, .backend = memory.Backend.fromIO(io_mem) };
+    const io_space = try memory.MemorySpace.init(alloc, io_seg_buf[0..]);
+
+    // Program space: byte-addressable view over Flash
```

Comment:
Flash accesses are a bit uglier now since I shoe-horned them into an interface that only does read8. I could add a read16 method to the vtable, and make them error out in the ram/io implementations maybe?